ledger-service/http-json/src/main/scala/com/digitalasset/http/domain.scala

```diff
@@ -252,32 +262,159 @@ package domain {
     final case class Offset(offset: HexString) extends domain.DeduplicationPeriod
   }
 
-  final case class CommandMeta(
+  final case class PbAny(typeUrl: String, value: Base64) {
+    import com.google.protobuf.any
+    def toLedgerApi: any.Any = any.Any(typeUrl, Base64 unwrap value)
+  }
+
+  final case class DisclosedContract[+TmplId, +LfV](
+      contractId: ContractId,
+      templateId: TmplId,
+      arguments: DisclosedContract.Arguments[LfV],
+      metadata: DisclosedContract.Metadata,
+  ) {
+    def toLedgerApi(implicit
+        TmplId: TmplId <:< ContractTypeId.Template.RequiredPkg,
+        LfV: LfV <:< lav1.value.Record,
+    ): lav1.commands.DisclosedContract =
+      lav1.commands.DisclosedContract(
+        templateId = Some(apiIdentifier(templateId)),
+        contractId = ContractId unwrap contractId,
+        arguments = arguments.toLedgerApi,
+        metadata = Some(metadata.toLedgerApi),
+      )
+  }
+
+  object DisclosedContract {
+    type LAV = DisclosedContract[ContractTypeId.Template.RequiredPkg, lav1.value.Record]
+
+    sealed abstract class Arguments[+LfV] extends Product with Serializable {
+      import lav1.commands.DisclosedContract.Arguments.{CreateArguments, CreateArgumentsBlob}
+      def toLedgerApi(implicit
+          LfV: LfV <:< lav1.value.Record
+      ): lav1.commands.DisclosedContract.Arguments = this match {
+        case Arguments.Blob(blob) => CreateArgumentsBlob(blob.toLedgerApi)
+        case Arguments.Record(payload) => CreateArguments(payload)
+      }
+    }
+    object Arguments {
+      final case class Blob(payloadBlob: PbAny) extends Arguments[Nothing]
+      final case class Record[+LfV](payload: LfV) extends Arguments[LfV]
+      private[http] val blobKey = "payloadBlob"
+      private[http] val recordKey = "payload"
+
+      implicit val covariant: Traverse[Arguments] = new Traverse[Arguments] {
+        override def traverseImpl[G[_]: Applicative, A, B](
+            fa: Arguments[A]
+        )(f: A => G[B]): G[Arguments[B]] =
+          fa match {
+            case b @ Blob(_) => Applicative[G] point b
+            case Record(payload) => f(payload) map (Record(_))
+          }
+      }
+    }
+
+    final case class Metadata(
+        createdAt: Time.Timestamp,
+        contractKeyHash: Option[Base64],
+        driverMetadata: Option[Base64],
+    ) {
+      import com.daml.api.util.TimestampConversion.{fromLf => timestampLfToProto}
+
+      def toLedgerApi: lav1.contract_metadata.ContractMetadata =
+        lav1.contract_metadata.ContractMetadata(
+          createdAt = Some(timestampLfToProto(createdAt)),
+          contractKeyHash = contractKeyHash.cata(Base64.unwrap, ByteString.EMPTY),
```

Comment:
I think I would suggest to keep the contract key hash base16. Seems similar to a contract id which is also base16. For the other ones I'm fine with base64.

Reply:
Included in this PR.